### ISSUE TITLE
Sik-Ho Tang | Brief Review -- RCNN: Recurrent Convolutional Neural Network for Object Recognition.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#98]
Sik-Ho Tang. [Brief Review — RCNN: Recurrent Convolutional Neural Network for Object Recognition](https://sh-tsang.medium.com/brief-review-rcnn-recurrent-convolutional-neural-network-for-object-recognition-8c9870d4879a).

**Comment by NorbertZheng** [comment]
## Overview
Recurrent Convolutional Neural Network for Object Recognition. RCNN, by Tsinghua University. 2015 CVPR, Over 1000 Citations.

Image Classification, CNN, RNN.

A Recurrent CNN (RCNN) is proposed for image classification, where **the convolutional layer is recurrently used for multiple times**.

**Comment by NorbertZheng** [comment]
## Recurrent CNN (RCNN)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/779534da-5bf8-4b01-9781-bcc0c09b3418)
RCNN Overall Network Architecture.

### Recurrent Convolutional Layer (RCL)
 - The key module of RCNN is the recurrent convolutional layer (RCL).

For a unit located at $(i, j)$ on the kth feature map in an RCL, its net input $z_{ijk}(t)$ at time step $t$ is given by:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/05780256-8062-465d-9370-9f85fbd45424)

The activity or state of this unit is a function of its net input:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0efd6382-4184-44f8-b1ae-cd2ce37d0dc5)

where f is the rectified linear activation ([ReLU](https://sh-tsang.medium.com/brief-review-rectified-linear-units-improve-restricted-boltzmann-machines-43724ede3ecf)) function:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/ad879bb9-a3ed-4fa6-ada0-5f3465352a3c)

and g is the local response normalization (LRN) function in [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160?source=post_page---------------------------):
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/bcd53c98-4d8d-4766-abe5-fe8a17e60ff4)

It is claimed that LRN is used for preventing the states from exploding. **Maybe we can replace LRN with BN or LN, etc.**

### RCNN
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/779534da-5bf8-4b01-9781-bcc0c09b3418)
RCNN Overall Network Architecture.

Left: An example with $T=3$. When $t=0$ only the feedforward input is present.

The final gradient of a shared weight is the sum of its gradients over all time steps.

Right: **To save computation, layer 1 is the standard feed-forward convolutional layer without recurrent connections, followed by max pooling.** On top of this, four RCLs are used with a max pooling layer in the middle. Both pooling operations have stride 2 and size 3.

The output of the fourth RCL follows a global max pooling, yielding a feature vector representing the image.

Finally a softmax layer is used to classify the feature vectors to $C$ categories whose output is given by:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/06dae627-fa29-4fa0-aa1c-8546072a2fcd)

The cross-entropy loss function is used for training.

If we unfold the recurrent connections for $T$ time steps, the model becomes a very deep feed-forward network with $4(T+1)+2$ parameterized layers, where $T+1$ is the depth of each RCL.

**Comment by NorbertZheng** [comment]
## Results
### CIFAR-10
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c82d9cef-69c6-4f87-aecb-e5e2f373fbc6)
Comparison with existing models on CIFAR-10.

Three models with different $K$’s were tested: RCNN-96, RCNN-128 and RCNN-160. The number of iterations was set to $3$.

All of them outperformed existing models such as [Maxout](https://towardsdatascience.com/review-maxout-network-image-classification-40ecd77f7ce4?source=post_page---------------------------) and [NIN](https://towardsdatascience.com/review-nin-network-in-network-image-classification-69e271e499ee?source=post_page---------------------------), and the performance was steadily improved with more features maps.

### CIFAR-100
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0af203a8-6bfb-4191-821d-9c9a364123eb)
Comparison with existing models on CIFAR-100.

Again RCNN-96 outperformed the state-of-the-art models with fewer parameters, and the performance kept improving by increasing $K$.

### MNIST
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e8159125-7df4-4a59-a9e4-c1c21d825b54)
Comparison with existing models on MNIST.

RCNN-64 outperformed other models using only 0.30 million parameters.

### SVHN
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0d7aa477-5960-4338-944c-2719da6d2b43)
Comparison with existing models on SVHN.

RCNN-128 had much fewer parameters than [NIN](https://towardsdatascience.com/review-nin-network-in-network-image-classification-69e271e499ee?source=post_page---------------------------) (1.19 million versus 1.98 million), and increasing $K$ kept improving the accuracy.

**Comment by NorbertZheng** [comment]
Similar idea is used in PolyInception Modules as in [PolyNet](https://towardsdatascience.com/review-polynet-2nd-runner-up-in-ilsvrc-2016-image-classification-8a1a941ce9ea), and [PolyNet](https://towardsdatascience.com/review-polynet-2nd-runner-up-in-ilsvrc-2016-image-classification-8a1a941ce9ea) got 2nd Runner Up in ILSVRC 2016 Image Classification.

**Comment by NorbertZheng** [comment]
## Reference
 - [2015 CVPR] [RCNN] [Recurrent Convolutional Neural Network for Object Recognition](https://www.cv-foundation.org/openaccess/content_cvpr_2015/papers/Liang_Recurrent_Convolutional_Neural_2015_CVPR_paper.pdf).
